### PR TITLE
Automated cherry pick of #123233: Fix daemonset to ensure old unhealthy pods are counted

### PR DESCRIPTION
**Pull request by ncdc**
Cherry pick of #123233 on release-1.28.

#123233: Fix daemonset to ensure old unhealthy pods are counted

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Ensure daemonset controller to count old unhealthy pods towards max unavailable budget
```

### CODE REVIEW

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/triage accepted
/priority important-longterm

**Comment by bridgetkromhout** [comment]
Cherry-pick to 1.31 not needed as the original commit merged early in the 1.31 release cycle, well before the 1.31 release was cut.

/cc kubernetes/release-managers

**Comment by liggitt** [comment]
I'm torn on whether this is a good candidate to cherry-pick. I'm trying to see when this issue got introduced... if this is fixing a regression that would keep someone from upgrading to 1.28+, it's a much clearer candidate for a cherry-pick. It looks like maybe in https://github.com/kubernetes/kubernetes/pull/96441/files#diff-f30e387a73b0d2fe9b55a074f9958bf1cae59724bbf297cf9f2466301af859c1 in 1.21?

We have to be *super* certain this won't regress controller behavior at all, and looking at the fix closely in https://github.com/kubernetes/kubernetes/pull/123233/files#r1787864007, I think it does regress how quickly the controller can clean up old unavailable pods.

**Comment by soltysh** [comment]
/hold

Jordan has a point, that we should put additional care before landing this PR

**Comment by liggitt** [comment]
I made a suggestion at https://github.com/kubernetes/kubernetes/pull/123233#discussion_r1787891756 with a possible alternate minimal fix that preserves old unavailable pod cleanup behavior.

Since this does look to be fixing a regression (albeit an old one from 1.21), I think I'd ack a backport with good unit tests for the scenario I described at https://github.com/kubernetes/kubernetes/pull/123233/files#r1787864007.

**Comment by bridgetkromhout** [comment]
@liggitt Thanks for looking at this. Can you clarify if you will approve this backport for merge after the merge of https://github.com/kubernetes/kubernetes/pull/127903 (and possible backport of it?) Thanks!

**Review by saschagrunert — APPROVED** [review]

**Comment by liggitt** [comment]
re-reviewing the change, I misread the logic and now no longer think this regresses the cleanup case - see https://github.com/kubernetes/kubernetes/pull/123233/files#r1795987494

/hold cancel

I would very cautiously lgtm this for backport
/lgtm
/approve

**Review by saschagrunert — APPROVED** [review]

**Comment by soltysh** [comment]
Looks like https://github.com/kubernetes/kubernetes/issues/119763
/retest

**Pull request merged**
